File: backend/internal/models/group.go
package models

import "time"

// GroupRole defines the permission level within a group.
type GroupRole string

const (
	RoleOwner  GroupRole = "OWNER"
	RoleEditor GroupRole = "EDITOR"
	RoleViewer GroupRole = "VIEWER"
)

// UserGroup defines a collection of users who share transactions and categories.
type UserGroup struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"createdAt"`
}

// CreateGroupRequest is used by Admins to manually create a new group.
type CreateGroupRequest struct {
	Name string `json:"name"`
}

// ManageGroupMembersRequest is used to add/remove users from a group.
type ManageGroupMembersRequest struct {
	UserIDs []string `json:"userIDs"`
}

// GroupMember represents a user with their role in a group.
type GroupMember struct {
	UserID   string    `json:"userId"`
	Username string    `json:"username"`
	Role     GroupRole `json:"role"`
	JoinedAt string    `json:"joinedAt"`
}

// GroupInfo is returned by GET /api/me/group (the user's current group context).
type GroupInfo struct {
	ID          string        `json:"id"`
	Name        string        `json:"name"`
	MemberCount int           `json:"memberCount"`
	Members     []GroupMember `json:"members"`
	MyRole      GroupRole     `json:"myRole"`
}

// ActivityLogEntry represents one activity record.
type ActivityLogEntry struct {
	ID         string `json:"id"`
	GroupID    string `json:"groupId"`
	UserID     string `json:"userId"`
	Username   string `json:"username"`
	Action     string `json:"action"`
	EntityType string `json:"entityType"`
	EntityID   string `json:"entityId"`
	Details    string `json:"details"`
	CreatedAt  string `json:"createdAt"`
}

// GroupInvite represents an invitation to join a group.
type GroupInvite struct {
	Code      string `json:"code"`
	GroupName string `json:"groupName"`
	CreatedBy string `json:"createdBy"`
	ExpiresAt string `json:"expiresAt"`
}

// CreateInviteResponse is the response after creating an invite.
type CreateInviteResponse struct {
	Code      string `json:"code"`
	ExpiresAt string `json:"expiresAt"`
}

// JoinGroupRequest is the payload for joining a group via invite code.
type JoinGroupRequest struct {
	InviteCode string `json:"inviteCode"`
}

// UpdateGroupNameRequest is the payload for PATCH /api/me/group.
type UpdateGroupNameRequest struct {
	Name string `json:"name"`
}

File: frontend/src/api/group.ts
import { apiFetch } from "./client";

export interface GroupMember {
  userId: string;
  username: string;
  role: "OWNER" | "EDITOR" | "VIEWER";
  joinedAt: string;
}

export interface GroupInfo {
  id: string;
  name: string;
  memberCount: number;
  members: GroupMember[];
  myRole: "OWNER" | "EDITOR" | "VIEWER";
}

export interface ActivityLogEntry {
  id: string;
  groupId: string;
  userId: string;
  username: string;
  action: string;
  entityType: string;
  entityId: string;
  details: string;
  createdAt: string;
}

export interface CreateInviteResponse {
  code: string;
  expiresAt: string;
}

export async function getMyGroup(): Promise<GroupInfo> {
  return apiFetch("/api/me/group");
}

export async function updateGroupName(name: string) {
  return apiFetch("/api/me/group", {
    method: "PATCH",
    body: JSON.stringify({ name }),
  });
}

export async function getActivityFeed(limit = 10): Promise<ActivityLogEntry[]> {
  return apiFetch(`/api/me/activity?limit=${limit}`);
}

export async function createInvite(): Promise<CreateInviteResponse> {
  return apiFetch("/api/me/group/invite", {
    method: "POST",
  });
}

export async function joinGroup(inviteCode: string) {
  return apiFetch("/api/me/group/join", {
    method: "POST",
    body: JSON.stringify({ inviteCode }),
  });
}

export async function leaveGroup() {
  return apiFetch("/api/me/group/leave", {
    method: "POST",
  });
}

// --- Multi-Group Support ---

export interface GroupSummary {
  id: string;
  name: string;
  memberCount: number;
  myRole: "OWNER" | "EDITOR" | "VIEWER";
}

export interface SwitchGroupResponse {
  token: string;
  groupId: string;
  groupName: string;
  groupRole: string;
}

export async function listMyGroups(): Promise<GroupSummary[]> {
  return apiFetch("/api/me/groups");
}

export async function createMyGroup(name: string): Promise<GroupSummary> {
  return apiFetch("/api/me/groups", {
    method: "POST",
    body: JSON.stringify({ name }),
  });
}

export async function switchGroup(groupId: string): Promise<SwitchGroupResponse> {
  return apiFetch("/api/me/switch-group", {
    method: "POST",
    body: JSON.stringify({ groupId }),
  });
}

export async function deleteMyGroup(groupId: string): Promise<void> {
  return apiFetch(`/api/me/groups/${groupId}`, {
    method: "DELETE",
  });
}
